src/event_filter.rs

```diff
@@ -0,0 +1,165 @@
+    /// of changes.
+    Modify(Utf8PathBuf),
+    /// A file is removed.
+    Remove(Utf8PathBuf),
+}
+
+/// Process the events contained in an [`Action`] into a list of [`FileEvent`]s.
+pub fn file_events_from_action(action: &Action) -> miette::Result<Vec<FileEvent>> {
+    // First, build up a map from paths to events tagged with that path.
+    // This will give us easy access to the events for a given path in this batch.
+    let mut events_by_path = HashMap::<Utf8PathBuf, Vec<&Event>>::new();
+    for event in action.events.iter() {
+        for tag in event.tags.iter() {
+            if let Tag::Path { path, .. } = tag {
+                let path = path.to_owned().try_into().into_diagnostic()?;
+                let entry = events_by_path.entry(path).or_insert_with(Vec::new);
+                entry.push(event);
+                // No need to look at the rest of the tags.
+                break;
+            }
+        }
+    }
+
+    let mut ret = Vec::new();
+
+    for (path, events) in events_by_path.iter() {
+        if path
+            .extension()
+            .map(|ext| !HASKELL_SOURCE_EXTENSIONS.contains(&ext))
+            .unwrap_or(true)
+        {
+            // If the path doesn't have a Haskell source extension, we don't need to process it.
+            // In the future, we'll want something more sophisticated here -- we'll need to reload
+            // for non-Haskell files or even run commands when non-Haskell files change -- but this
+            // is fine for a first pass.
+            continue;
+        }
+
+        let mut exists = false;
+        let mut created = false;
+        let mut modified = false;
+        let mut removed = false;
+        let mut renamed = false;
+        for event in events {
+            for tag in event.tags.iter() {
+                match tag {
+                    Tag::Path { path: _, file_type } => {
+                        exists = file_type.is_some();
+                    }
+                    Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Name(_))) => {
+                        renamed = true;
+                    }
+                    Tag::FileEventKind(FileEventKind::Modify(ModifyKind::Data(_))) => {
+                        modified = true;
+                    }
+                    Tag::FileEventKind(FileEventKind::Create(CreateKind::File)) => {
+                        created = true;
+                    }
+                    Tag::FileEventKind(FileEventKind::Remove(RemoveKind::File)) => {
+                        removed = true;
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        // Write existing file from Vim:    exists, renamed, created, modified
+        // Write existing file from VSCode: exists,          created, modified
+        // Create a new file:               exists,          created, modified
+        // `mv`:                            exists, renamed, created, modified
```

Comment:
Is this for the old filename or the new filename?  The reason I ask is that my understanding is that `exists` would be `false` for the old filename:

```suggestion
        // `mv`:                           !exists, renamed, created, modified
```

Reply:
Meant the new filename but it wasn't right. Fixed the comment.